File: web/src/main/java/jumpstart/web/mixins/Tabber.java
/**
 * A mixin that watches that watches every link (<a>) contained in the given element. 
 * When one is clicked it sets the class of the link to "active" and the class of the others to "".
 */

package jumpstart.web.mixins;

import org.apache.tapestry5.BindingConstants;
import org.apache.tapestry5.ClientElement;
import org.apache.tapestry5.RenderSupport;
import org.apache.tapestry5.annotations.AfterRender;
import org.apache.tapestry5.annotations.IncludeJavaScriptLibrary;
import org.apache.tapestry5.annotations.InjectContainer;
import org.apache.tapestry5.annotations.Parameter;
import org.apache.tapestry5.ioc.annotations.Inject;

//This annotation tells Tapestry to declare the js file in the page so that the browser will pull it in.
@IncludeJavaScriptLibrary("tabber.js")
public class Tabber {

	@Parameter(defaultPrefix = BindingConstants.PROP)
	private Integer activeTabNum;

	@Inject
	private RenderSupport renderSupport;

	@InjectContainer
	private ClientElement element;

	@AfterRender
	public void afterRender() {
		// Tell Tapestry to add some javascript that instantiates a Tabber for the element we're mixing into.
		// Tapestry will put it at the end of the page in a section that runs once the DOM has been loaded.
		// The Tabber class it refers to is NOT THIS class - it is actually the one defined in tabber.js.
		renderSupport.addScript(String.format("new Tabber('%s', '%d');", element.getClientId(), activeTabNum));
	}

}

File: web/src/main/java/jumpstart/web/mixins/tabber.js
// From http://wiki.apache.org/tapestry/Tapestry5AndJavaScriptExplained

/*
 * A mixin that watches that watches every link (<a>) contained in the given element.
 * When one is clicked it sets the class of the link to "active" and the class of the others to "".
 */

var Tabber = Class.create();
var links = null;

Tabber.prototype = {
		
        initialize: function(elementId, activeLinkNum) {
//	alert('initializeactiveLinkNum = ' + activeLinkNum);
			var me = this;
			links = $$('#' + elementId + ' a');

			links.each(function(link) {
				Event.observe(link, 'click', me.doLinkClicked.bindAsEventListener(me));
			});

			var target = links[activeLinkNum];
			this.makeActive(target);
		},
        
		doLinkClicked: function(evt) {
//	alert('doTabClicked: evt = ' + evt);
			var target = Event.element(evt);
			this.makeActive(target);
			
			return false;
		},
        
		makeActive: function(element) {
//	alert('makeActive: element =' + element);
			
			links.each(function(link) {
				if (link == element) {
//					alert('Same');
					link.className = 'active';
				}
				else {
//					alert('Different');
					link.className = '';
				}
			});
		}
		
		// Where should this IE stuff go?
//		$(document).ready(function(){
//			  $("ul.tabs li").hover(
//			    function() { $(this).addClass("iehover"); },
//			    function() { $(this).removeClass("iehover"); }
//			  );
//		});

}